extensions/amp-video/0.1/amp-video.js

```diff
@@ -204,38 +243,131 @@ class AmpVideo extends AMP.BaseElement {
-    }
+    const viewer = Services.viewerForDoc(this.getAmpDoc());
 
-    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT, this.video_,
+    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT,
+        dev().assertElement(this.video_),
         /* opt_removeMissingAttrs */ true);
 
-    this.getRealChildNodes().forEach(child => {
-      // Skip the video we already added to the element.
-      if (this.video_ === child) {
-        return;
-      }
-      if (child.getAttribute && child.getAttribute('src')) {
-        assertHttpsUrl(child.getAttribute('src'),
-            dev().assertElement(child));
+    if (viewer.getVisibilityState() == VisibilityState.PRERENDER) {
+      if (!this.element.hasAttribute('preload')) {
+        this.video_.setAttribute('preload', 'auto');
       }
-      this.video_.appendChild(child);
-    });
+      this.propagatePrerenderSources_();
+      viewer.whenFirstVisible().then(() => {
+        this.propagateLayoutChildren_();
+      });
+    } else {
+      this.propagatePrerenderSources_();
+      this.propagateLayoutChildren_();
+    }
 
     // loadPromise for media elements listens to `loadstart`
     return this.loadPromise(this.video_).then(() => {
       this.element.dispatchCustomEvent(VideoEvents.LOAD);
     });
   }
 
+  /**
+   * @private
+   * Propagate sources that are cached by the CDN.
+   */
+  propagatePrerenderSources_() {
+    dev().assert(this.video_);
+
+    const sources = toArray(childElementsByTag(this.element, 'source'));
+
+    // if the `src` of `amp-video` itself is cached, move it to <source>
+    if (this.element.hasAttribute('src') && this.isCachedByCDN_(this.element)) {
+      const src = this.element.getAttribute('src');
+      const ampOrigSrc = this.element.getAttribute('amp-orig-src');
+      assertHttpsUrl(src, this.element);
+      const srcSource = this.element.ownerDocument.createElement('source');
+      srcSource.setAttribute('src', src);
+      srcSource.setAttribute('amp-orig-src', ampOrigSrc);
+      sources.unshift(srcSource);
+    }
+
+    sources.forEach(source => {
+      if (this.isCachedByCDN_(source)) {
+        this.video_.appendChild(source);
+      }
+    });
+  }
+
+  /**
+   * Propagate origin sources and tracks
+   * @private
+   */
+  propagateLayoutChildren_() {
+    dev().assert(this.video_);
+
+    const sources = toArray(childElementsByTag(this.element, 'source'));
+
+    // If the `src` of `amp-video` itself is NOT cached, set it on video
+    if (this.element.hasAttribute('src') &&
+        !this.isCachedByCDN_(this.element)) {
+      assertHttpsUrl(this.element.getAttribute('src'), this.element);
+      this.propagateAttributes(['src'], dev().assertElement(this.video_));
+    }
+
+    sources.forEach(source => {
+      // Cached sources should have been moved already.
+      dev().assert(!this.isCachedByCDN_(source));
+      assertHttpsUrl(source.getAttribute('src'), source);
+      this.video_.appendChild(source);
```

Comment:
Do we need to retain original order?

Reply:
that's a tough one, given the original:
```
<source type=webm>
<source type=mp4>
```
where only `mp4` source is cached:

What would perform better? cached `mp4` or origin `webm`? for a fast origin, maybe `webm` but we don't know that. Not having cached sources first also means we may throw out all the preloading browser has done.

I believe having cached ones first is the right choice unless proven otherwise.

For the moment, it does not matter since cache rewrites Url all sources regardless of type so we never get into a mix of cached and non-cached sources (the rewritten url may 404 for not supported types which effectively preserves the order)

Reply:
I think a nice property of preserving the order is that the cached page performance cannot be worse than the original page. While we would lose some potential gains, we are "guaranteed" to never make things worse.
I think for HLS followed by MP4, there is a good chance that the cached mp4 performs worse than the non-cached HLS.

Having said that, this will only matter in fallback cases (e.g., cache miss/video too large to be cached/video-type not supported) so hopefully either way it's fine, and maybe we should just go with whatever is more robust (i.e., less likely to break the (auto)playing experience).

The same order question applies to adding the amp-orig-src sources.

Reply:
`amp-orig-src` are always added right after the cached source so we are good there. Based on our chat, given every Url gets overridden anyway, effectively order of sources is preserved and  every `<source origin>` is replaced  with `<source cached> <source origin>` pair in place.